Add option to build SHA-512 without SHA-384

mpg opened the pull request:
This adds a config.h option `MBEDTLS_SHA512_NO_SHA384` that allows to build with SHA-512 enabled but not SHA-384, for people who don't need SHA-384 and want to save on code size.

This is very comparable to https://github.com/ARMmbed/mbedtls-restricted/pull/621 (which should be side-ported).

Contrary to SHA-256 however, it seems desirable to also have a symmetrical option for people who only need SHA-384 but not SHA-512. In order to keep this PR small and allow it to move forward quickly, it was not included but can be done later.

Reviewers, please double-check if I didn't miss anything to do with PSA as I'm less familiar with that part.

Size, building with `CFLAGS='-Os -mcpu=cortex-m0plus -mthumb' CC=arm-none-eabi-gcc` (ARM-GCC 7.3) after running `scripts/config.pl baremetal` (don't pay attention to the absolute values as this is a full config, only the difference is interesting):

|  | Default | Without selftest |
| --- | --- | --- |
| `libmbedcrypto.a` with SHA-384 |  228769 | 192320 |
| `libmbedcrypto.a` without SHA-384 | 228353 | 192128 |
| gain in Bytes | 416 | 192 |

**Note:** If #327 is merged first, this PR will need to be modified to add a dependency on `!MBEDTLS_SHA512_NO_SHA384` to some test cases added by #327.

gilles-peskine-arm commented:
Please extend `tests/scripts/depends-hashes.pl` to support the new options.

mpg commented:
Note: an extension of `depends-hashes.pl` for a similar option has been done in cousin PR https://github.com/ARMmbed/mbedtls-restricted/pull/621 and would be pretty straightforward to adapt here.

mpg commented:
@gilles-peskine-arm I rebased in order to resolve conflicts (previous state is at https://github.com/mpg/mbed-crypto/tree/sha512-no-sha384-1) and extended `depends-hashes.pl` as you suggested. I think this is now ready for review again.

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -2063,8 +2063,10 @@ static const mbedtls_md_info_t *mbedtls_md_info_from_psa( psa_algorithm_t alg )
 #endif
 #if defined(MBEDTLS_SHA512_C)
+#if !defined(MBEDTLS_SHA512_NO_SHA384)
         case PSA_ALG_SHA_384:
```
Missing: `case PSA_ALG_SHA_384` in `psa_hash_setup` (the other cases have negligible impact)

gilles-peskine-arm commented on library/sha512.c:
```diff
@@ -150,6 +150,9 @@ int mbedtls_sha512_starts_ret( mbedtls_sha512_context *ctx, int is384 )
     }
     else
```
Missing above: `SHA512_VALIDATE_RET` should not accept `is384 == 1` if SHA384 is disabled.

gilles-peskine-arm commented on library/sha512.c:
```diff
@@ -568,6 +578,10 @@ static const unsigned char sha512_test_sum[6][64] =
       0x4E, 0xAD, 0xB2, 0x17, 0xAD, 0x8C, 0xC0, 0x9B }
 };
 
+#define SHA512_TEST_SUM_N \
```
Minor: I'd prefer to write `ARRAY_LENGTH(sha512_test_sum)` with a generic `ARRAY_LENGTH` macro so that it's evident at the point of use that we're looking at the length of the relevant array.

gilles-peskine-arm commented on tests/suites/test_suite_pkparse.data:
```diff
@@ -651,147 +651,147 @@ depends_on:MBEDTLS_DES_C:MBEDTLS_SHA256_C:MBEDTLS_PKCS5_C
 
 Parse RSA Key #75 (PKCS#8 encrypted v2 PBKDF2 3DES hmacWithSHA384)
-depends_on:MBEDTLS_DES_C:MBEDTLS_SHA512_C:MBEDTLS_PEM_PARSE_C:MBEDTLS_PKCS5_C:MBEDTLS_CIPHER_MODE_CBC
+depends_on:MBEDTLS_DES_C:MBEDTLS_SHA512_C:MBEDTLS_PEM_PARSE_C:MBEDTLS_PKCS5_C:MBEDTLS_CIPHER_MODE_CBC:!MBEDTLS_SHA512_NO_SHA384
```
Minor: I think it would be easier to read if `!MBEDTLS_SHA512_NO_SHA384` was consecutive to `MBEDTLS_SHA512_C`.

gilles-peskine-arm commented on tests/suites/test_suite_rsa.data:
```diff
@@ -23,11 +23,11 @@ depends_on:MBEDTLS_SHA256_C:MBEDTLS_PKCS1_V15
 
 RSA PKCS1 Verify v1.5 CAVS #6
-depends_on:MBEDTLS_SHA512_C:MBEDTLS_PKCS1_V15
+depends_on:MBEDTLS_SHA512_C:MBEDTLS_PKCS1_V15:!MBEDTLS_SHA512_NO_SHA384:!MBEDTLS_SHA512_NO_SHA384
```
Minor: once is enough.

piotr-now commented on include/mbedtls/md_internal.h:
```diff
@@ -79,7 +79,9 @@ extern const mbedtls_md_info_t mbedtls_sha224_info;
 #if defined(MBEDTLS_SHA512_C)
+#if !defined(MBEDTLS_SHA512_NO_SHA384)
 extern const mbedtls_md_info_t mbedtls_sha384_info;
+#endif
```
There is many places when you use this sentence with double negation "`#if !defined(MBEDTLS_SHA512_NO_SHA384)`" an it makes riding it is not very friendly. This is a second advantage to use option like `MBEDTLS_SHA512_WITH_SHA384`

mpg commented on include/mbedtls/config.h:
```diff
@@ -1026,6 +1026,16 @@
+ * Uncomment to disable SHA-384
+ */
+//#define MBEDTLS_SHA512_NO_SHA384
+
```
Regarding `NO` vs `WITH`, I agree in principle, but unfortunately there's a backwards-compatibility issue. Currently people who use a custom `config.h` can just define `MBEDTLS_SHA512_C` and they get SHA-384 as well. If make SHA-384 available only if `MBEDTLS_SHA512_WITH_SHA384` is defined as well, it means they have to change their code (namely, their `config.h`) when updating Mbed TLS, which one thing we promised they won't have to do except on major version upgrades.

So in Mbed TLS 3.0.0 I agree with should have a `WITH` option rather than a `NO` option. But in the meantime if we want to merge for a 2.x release, I'm afraid a `NO` option is our only choice.

What we could to however, is to translate it to a `WITH` option internally, so that we at least avoid the readability issue inside the `.c` file.

piotr-now replied:
It makes sense to me. Thanks for explanation

mpg commented on include/mbedtls/sha512.h:
```diff
@@ -59,8 +59,10 @@ typedef struct mbedtls_sha512_context
     unsigned char buffer[128];  /*!< The data block being processed. */
+#if !defined(MBEDTLS_SHA512_NO_SHA384)
     int is384;                  /*!< Determines which function to use:
                                      0: Use SHA-512, or 1: Use SHA-384. */
```
That would be an ABI change, so I'd rather not to it here.

mpg commented:
I just rebased on current development and force-pushed in order to resolve conflicts. The previous state of the branch was saved at https://github.com/mpg/mbed-crypto/tree/sha512-no-sha384-2

 I'll push more commits in order to address review comments later.

mpg commented:
@gilles-peskine-arm @piotr-now I believe I addressed your comments, either by changing the code as suggested, or commenting why I'd rather not. This is ready for you to review again.

mpg commented on include/mbedtls/sha512.h:
```diff
@@ -101,7 +103,11 @@ void mbedtls_sha512_clone( mbedtls_sha512_context *dst,
- *                 either \c for SHA-512, or \c 1 for SHA-384.
+ *                 either \c 0 for SHA-512, or \c 1 for SHA-384.
+ *
+ * \note           When MBEDTLS_SHA512_NO_SHA384 is defined, \p is384 must be
```
I assume you mean `#MBEDTLS_SHA512_NO_SHA384`? Or is there a subtlety I'm missing?

gilles-peskine-arm replied:
I meant `\c`, because when `MBEDTLS_SHA512_NO_SHA384` is not defined, the `#` version would render with the `#`.

mpg replied:
Oh, good point. Never noticed that.

mpg commented on library/psa_crypto.c:
```diff
@@ -2517,8 +2529,10 @@ static size_t psa_get_hash_block_size( psa_algorithm_t alg )
             return( 64 );
         case PSA_ALG_SHA_256:
             return( 64 );
+#if !defined(MBEDTLS_SHA512_NO_SHA384)
```
I'm not sure I understand what the inconsistency is. Could you clarify?

gilles-peskine-arm replied:
To be consistent, `case PSA_ALG_SHA_512` should be under `#if defined(MBEDTLS_SHA512_C)`, etc.

mpg commented:
@gilles-peskine-arm Thanks for your feedback, and clarifying when I didn't immediately understand it. I believe I've addressed all of it so feel free to review again!

mpg commented:
The CI passes all tests except the ones involving Mbed OS, which is a known issue independent from this PR, so it's as good as a pass as far as merging this PR is concerned.

gilles-peskine-arm commented on library/sha512.c:
```diff
@@ -587,7 +587,7 @@ static const unsigned char sha512_test_sum[][64] =
 };
 
-#define ARRAY_LEN(a)    ( sizeof( a ) / sizeof( a[0] ) )
+#define ARRAY_LENGTH(a)    ( sizeof( a ) / sizeof( a[0] ) )
```
```suggestion
#define ARRAY_LENGTH( a )    ( sizeof( a ) / sizeof( ( a )[0] ) )
```

mpg commented:
@gilles-peskine-arm Thanks for catching that issue. I fixed it, this should now be ready for re-review.

@piotr-now You reviewed and approved a previous version of this PR, do you think you'd be able to review it again? Thanks!

piotr-now commented:
LGTM

mpg commented:
The CI passes all tests except for the ones relying on mbed-os, which is a know issue with mbed-os, unrelated to this PR, so it's a s good as a pass.
